Problem: No type-safe alternatives when polling or needing a reference to a socket

gummif opened the pull request:
Solution: Introduce a socket_ref that is a non-owning nullable reference to a socket

This is a quite big PR with a few small additional types to improve type safety. The API is intentionally minimal, but can be extended at later points. Some design decision:

1. `socket_ref` is a trivially copyable type with the same layout as `void*`, making it a step towards typesafe polling
2. `socket_ref` is provides comparison and hashing, making it suitable for containers like `std::map` or `std::unordered_set`. In addition to being a non-owning reference, makes the use of `void*` obsolete (for C++11).
3. A `handle()` function is provided to get the underlying `void*` (solving https://github.com/zeromq/cppzmq/issues/140), since `native_handle()` implies an operating system native socket handle, the `handle()` name was chosen instead
3. A function taking void* is quite greedy, so to construct `socket_ref` from an existing handle one must use the `zmq::from_handle` utility to be explicit
4. Flag enums for safer `send` and `recv`
5. Result structs from `send` and `recv` removing ambiguity on EAGAIN (solving https://github.com/zeromq/cppzmq/issues/84 and https://github.com/zeromq/cppzmq/issues/61)

sigiesec commented on zmq.hpp:
```diff
@@ -634,7 +634,146 @@ enum class socket_type : int
+
+// A non-owning nullable reference to a socket.
+// The reference is invalidated on socket close or destruction.
+class socket_ref
```
I am a bit confused on the relationship of socket_ref and socket. So now socket_ref has new send/recv variants, which are not present on socket. This is somewhat confusing IMO.

However, I am not sure what I would expect from the API of a socket_ref. Wouldn't one want to have the same operations as on socket?

That would lead to quite a lot of duplication though.

Wouldn't it be easier and implicitly consistent to use the same class for an owning and a non-owning socket, or at least the same base class (or base class template)?

gummif replied:
Yes, the API is supposed to be mostly the same except for the `close` function. `socket_t` also has the burden of pre-C++11 support (is dropping that support on the horizon?), but I will see if I can refactor the common functionality into a base class.

gummif commented on zmq.hpp:
```diff
@@ -634,7 +634,146 @@ enum class socket_type : int
+{
+    return a.handle() < b.handle();
+}
+inline bool operator>(socket_ref a, socket_ref b) noexcept
```
For completeness, but arguable not very useful. Would you like to see only < implemented?

gummif commented:
I have moved common functionality into `socket_base`, made `socket_ref` compile with C++98, replaced uses of `void*` to `socket_ref` and `socket_t` where possible, and added a `handle()` function to `message_t` removing the need for friend. The `send` and `recv` variants previously in the PR will be the subject of another PR.

sigiesec commented:
Thanks for the update, I was AFAIK for the last days. I hope I can give it a look later today.

sigiesec commented:
Sorry it took so long, I now merged this. Thanks again for the PR and the followup changes!
